### PR TITLE
[Expression]add more functions

### PR DESCRIPTION
**Pull request by feich-ms**
Add more expression functions including R1 and R2

### CODE REVIEW

File: libraries/botbuilder-expression/src/buildInFunction.ts [hunk 1]
```diff
@@ -970,6 +1137,76 @@ export class BuiltInFunctions {
...
+                ExpressionType.Skip,
+                BuiltInFunctions.Skip,
+                ReturnType.Object,
+                (expression: Expression): void => BuiltInFunctions.ValidateOrder(expression, [], ReturnType.Object, ReturnType.Number),
+            ),
+            new ExpressionEvaluator(
+                ExpressionType.Take,
+                BuiltInFunctions.Take,
+                ReturnType.Object,
+                (expression: Expression): void => BuiltInFunctions.ValidateArityAndAnyType(expression, 2, 2, ReturnType.String, ReturnType.Object, ReturnType.Number),
```
Comment (Danieladu) [inline libraries/botbuilder-expression/src/buildInFunction.ts]: maybe `ValidateOrder` is more reasonable?

File: libraries/botbuilder-expression/src/buildInFunction.ts [hunk 2]
```diff
@@ -1068,6 +1305,48 @@ export class BuiltInFunctions {
...
+            new ExpressionEvaluator(
+                ExpressionType.CountWord,
+                BuiltInFunctions.Apply((args: ReadonlyArray<any>) => args[0].trim().split(/\s+/).length, BuiltInFunctions.VerifyString),
+                ReturnType.Number,
+                BuiltInFunctions.ValidateUnaryString
+            ),
+            new ExpressionEvaluator(
+                ExpressionType.AddOrdinal,
+                BuiltInFunctions.Apply((args: ReadonlyArray<any>) => this.AddOrdinal(args[0]), BuiltInFunctions.VerifyInteger),
+                ReturnType.Number,
```
Comment (Danieladu) [inline libraries/botbuilder-expression/src/buildInFunction.ts]: return type of `AddOrdinal` is string?

File: libraries/botbuilder-expression/src/buildInFunction.ts [hunk 3]
```diff
@@ -1068,6 +1305,48 @@ export class BuiltInFunctions {
...
+                (expression: Expression): void => BuiltInFunctions.ValidateArityAndAnyType(expression, 1, 1, ReturnType.Number),
+            ),
+            new ExpressionEvaluator(
+                ExpressionType.Guid,
+                BuiltInFunctions.Apply(() => BuiltInFunctions.newGuid()),
+                ReturnType.String,
+                (expression: Expression): void => BuiltInFunctions.ValidateArityAndAnyType(expression, 0, 0, ReturnType.String),
```
Comment (Danieladu) [inline libraries/botbuilder-expression/src/buildInFunction.ts]: there is no need to write `ReturnType.String`

File: libraries/botbuilder-expression/src/buildInFunction.ts (hunk 1 above)
Comment (feich-ms) [reply libraries/botbuilder-expression/src/buildInFunction.ts]: yes. fixed

File: libraries/botbuilder-expression/src/buildInFunction.ts (hunk 2 above)
Comment (feich-ms) [reply libraries/botbuilder-expression/src/buildInFunction.ts]: my mistake, fixed.

File: libraries/botbuilder-expression/src/buildInFunction.ts (hunk 3 above)
Comment (feich-ms) [reply libraries/botbuilder-expression/src/buildInFunction.ts]: Right as there is no param. Fixed

File: libraries/botbuilder-expression-parser/tests/expression.test.js [hunk 4]
```diff
@@ -91,6 +92,27 @@ const dataSource = [
...
+  ["addOrdinal(11 + 10)", "21st"],
+  ["addOrdinal(11 + 11)", "22nd"],
+  ["addOrdinal(11 + 12)", "23rd"],
+  ["addOrdinal(11 + 13)", "24th"],
+  ["addOrdinal(-1)", "-1"],//original string value
+  ["count(guid())", 36],
+  ["guid().indexOf('-')", 8],
+  ["indexOf(guid(), '-')", 8],
+  ["guid().lastIndexOf('-')", 23],
+  ["lastIndexOf(guid(), '-')", 23],
```
Comment (Danieladu) [inline libraries/botbuilder-expression-parser/tests/expression.test.js]: add `indexof` and `lastIndexOf` tests that return -1

Comment (feich-ms) [reply libraries/botbuilder-expression-parser/tests/expression.test.js]: fixed.

File: libraries/botbuilder-expression/src/buildInFunction.ts [hunk 5]
```diff
@@ -1293,6 +1626,62 @@ export class BuiltInFunctions {
...
+                BuiltInFunctions.ValidateUnary),
+            new ExpressionEvaluator(
+                ExpressionType.Binary,
+                BuiltInFunctions.Apply((args: ReadonlyArray<any>) => this.ToBinary(args[0]), BuiltInFunctions.VerifyString),
+                ReturnType.String,
+                BuiltInFunctions.ValidateUnary),
+            new ExpressionEvaluator(
+                ExpressionType.DataUri,
+                BuiltInFunctions.Apply(
+                    (args: Readonly<any>) => 'data:text/plain;charset=utf-8;base64,'.concat(Buffer.from(args[0]).toString('base64')), BuiltInFunctions.VerifyString),
```
Comment (Danieladu) [inline libraries/botbuilder-expression/src/buildInFunction.ts]: personal opinion， is there a exist library to handle `datauri`, `dataUriToBinary` and `dataUriToString`. Maybe we can use a library to do more param checks. can [datauri](https://www.npmjs.com/package/datauri) help?

File: libraries/botbuilder-expression/src/buildInFunction.ts [hunk 6]
```diff
@@ -970,6 +1137,76 @@ export class BuiltInFunctions {
                     BuiltInFunctions.VerifyContainer),
                 ReturnType.Number,
                 BuiltInFunctions.ValidateUnary),
+            new ExpressionEvaluator(
+                ExpressionType.Range,
+                BuiltInFunctions.ApplyWithError(
+                    (args: ReadonlyArray<any>) => {
+                        let error: string;
+                        if (args[1] < 1) {
```
Comment (cosmicshuai) [inline libraries/botbuilder-expression/src/buildInFunction.ts]: should the condition be args[1] <= 0, according to the error messgae

File: libraries/botbuilder-expression/src/buildInFunction.ts (hunk 5 above)
Comment (feich-ms) [reply libraries/botbuilder-expression/src/buildInFunction.ts]: datauri package accept a file path as input which is not consistent with the one in WDL doc https://docs.microsoft.com/en-us/azure/logic-apps/workflow-definition-language-functions-reference#dataUri.

File: libraries/botbuilder-expression/src/buildInFunction.ts [hunk 7]
```diff
@@ -970,6 +1137,76 @@ export class BuiltInFunctions {
...
+                            error = 'Second paramter must be more than zero';
+                        }
+
+                        let num: number;
+                        let result: number[] = [];
+                        for (num = 0; num < args[1]; num++) {
+                            result = result.concat(Number(args[0]) + num);
```
Comment (Danieladu) [inline libraries/botbuilder-expression/src/buildInFunction.ts]: This method can be simplified:
```
var length: number = 4;
var start: number = 1;
var s: number[] = [...Array(length).keys()].map(e => e + start);
```

**Pull request merged**